lib/spack/spack/binary_distribution.py

```diff
@@ -631,3 +708,114 @@ def get_keys(install=False, trust=False, force=False):
                 else:
                     tty.msg('Will not add this key to trusted keys.'
                             'Use -t to install all downloaded keys')
+
+
+def read_from_url(file_uri):
+    file_contents = None
+
+    try:
+        url = urlopen(file_uri)
+    except URLError as url_err:
+        msg = 'Unable to open url {0} due to {1}'.format(
+            file_uri, url_err.message)
+        raise spack.error.SpackError(msg)
+    except Exception as expn:
+        msg = 'Error getting file contents at {0} due to {1}'.format(
+            file_uri, expn.message)
+        raise spack.error.SpackError(msg)
+
+    file_contents = url.read()
+
+    if not file_contents:
+        msg = 'Unable to read file contents at {0}'.format(file_uri)
+        raise spack.error.SpackError(msg)
+
+    return file_contents
+
+
+def needs_rebuild(spec, mirror_url, buildcache_index):
+    pkg_name = spec.name
+    pkg_version = spec.version
+
+    tty.msg('Checking {0}-{1}'.format(pkg_name, pkg_version))
+
+    spec.concretize()
```

Comment:
It's a little odd that this function does the concretizing -- I think it's more natural if the caller does it, especially since this modifies the `spec` passed in.